commands/previewPush.go

```diff
@@ -230,12 +264,105 @@ DomainLoop:
+			if !IsDomainManagedByRegistrar(cfg, domain, registrarName) {
+				fmt.Printf("Removing from provider %s: domain %s\n", registrarName, domain)
+				*totalCorrections += 1
+				if domainRemover, ok := registrar.(providers.DomainRemover); ok && push {
+					err := domainRemover.EnsureDomainAbsent(domain)
+					if err != nil {
+						out.Errorf("Error deleting domain: %s\n", err)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func DeleteUnmanagedZones(cfg *models.DNSConfig, createdProviders map[string]providers.DNSServiceProvider, push bool, out printer.CLI, totalCorrections *int) error {
+	fmt.Printf("Checking Zone Removal:\n")
+	for providerName, provider := range createdProviders {
+		zoneLister, ok := provider.(providers.ZoneLister)
+		if !ok {
+			out.Warnf("--purge-unmanaged-zones not implemented: provider %s\n", providerName)
+			continue
+		}
+		deployedZones, err := zoneLister.ListZones()
+		if err != nil {
+			return fmt.Errorf("failed ListZones for provider %s: %w\n", providerName, err)
+		}
+		for _, zone := range deployedZones {
+			if !IsZoneManagedByProvider(cfg, zone, providerName) {
+				fmt.Printf("Removing from provider %s: zone %s\n", providerName, zone)
+				*totalCorrections += 1
+				if zoneRemover, ok := provider.(providers.ZoneRemover); ok && push {
+					err := zoneRemover.EnsureZoneAbsent(zone)
+					if err != nil {
+						out.Errorf("Error deleting zone: %s\n", err)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func GetDomainCfg(cfg *models.DNSConfig, domain string) *models.DomainConfig {
+	for _, domainCfg := range cfg.Domains {
+		if domainCfg.Name == domain {
+			return domainCfg
+		}
+	}
+	return nil
+}
+
+func IsDomainManagedByRegistrar(cfg *models.DNSConfig, zone string, registrarName string) bool {
+	domainCfg := GetDomainCfg(cfg, zone)
+
+	if domainCfg == nil || domainCfg.RegistrarName != registrarName {
+		return false
+	}
+	return true
+}
+
+func IsZoneManagedByProvider(cfg *models.DNSConfig, zone string, dnsProviderName string) bool {
+	domainCfg := GetDomainCfg(cfg, zone)
+
+	if domainCfg == nil {
+		return false
+	}
+	for managedProviderName, _ := range domainCfg.DNSProviderNames {
+		if managedProviderName == dnsProviderName {
+			return true
+		}
+	}
+	return false
+}
+
+type ProviderState struct {
+	createdRegistrars   map[string]providers.Registrar
+	createdDnsProviders map[string]providers.DNSServiceProvider
+	notifier            notifications.Notifier
+	err                 error
+}
+
 // InitializeProviders takes (fully processed) configuration and instantiates all providers and returns them.
-func InitializeProviders(cfg *models.DNSConfig, providerConfigs map[string]map[string]string, notifyFlag bool) (notify notifications.Notifier, err error) {
+func InitializeProviders(cfg *models.DNSConfig, providerConfigs map[string]map[string]string, notifyFlag bool) ProviderState {
 	var notificationCfg map[string]string
-	defer func() {
-		notify = notifications.Init(notificationCfg)
-	}()
```

Comment:
Using this (with `notify` changed to `result.notifier`)  did result in the returned `providerState.notifier` being nil. 
I don't know why. 
To fix this, I have moved the `notifications.Init` out from the defer to before the (successful) return.

Is this okay?